--- a/ATmegaBOOT_168.c	2016-04-28 12:31:34.681987201 +0800
+++ b/ATmegaBOOT_168.c	2016-04-28 15:38:54.474458903 +0800
@@ -277,15 +276,18 @@
 	uint16_t w;
 
 #ifdef WATCHDOG_MODS
+
 	ch = MCUSR;
 	MCUSR = 0;
 
 	WDTCSR |= _BV(WDCE) | _BV(WDE);
 	WDTCSR = 0;
-
-	// Check if the WDT was used to reset, in which case we dont bootload and skip straight to the code. woot.
-	if (! (ch &  _BV(EXTRF))) // if its a not an external reset...
-		app_start();  // skip bootloader
+	
+    if (ch & (_BV(BORF) | _BV(PORF))) {
+		// After a power-on reset skip the bootloader and jump straight to sketch if one exists.	
+        if (pgm_read_word(0) != 0xFFFF) 
+	    	app_start();
+    }
 #else
 	asm volatile("nop\n\t");
 #endif
@@ -507,11 +509,6 @@
 	/* Leave programming mode  */
 	else if(ch=='Q') {
 		nothing_response();
-#ifdef WATCHDOG_MODS
-		// autoreset via watchdog (sneaky!)
-		WDTCSR = _BV(WDE);
-		while (1); // 16 ms
-#endif
 	}
 
 
